src/main/java/org/jitsi/srtp/crypto/OpenSslAesCipherSpi.java

```diff
@@ -0,0 +1,726 @@
+    {
+        AlgorithmParameterSpec spec = null;
+        if (params != null)
+        {
+            try
+            {
+                if (cipherMode == CipherMode.GCM)
+                {
+                    spec = params.getParameterSpec(GCMParameterSpec.class);
+                }
+                else
+                {
+                    spec = params.getParameterSpec(IvParameterSpec.class);
+                }
+            }
+            catch (InvalidParameterSpecException e)
+            {
+                throw new InvalidAlgorithmParameterException(e);
+            }
+        }
+        engineInit(opmode, key, spec, random);
+    }
+
+    @Override
+    protected byte[] engineUpdate(byte[] input, int inputOffset, int inputLen)
+    {
+        int bufNeeded = getOutputSize(inputLen, false);
+        byte[] buf = new byte[bufNeeded];
+        try
+        {
+            int len = engineUpdate(input, inputOffset, inputLen, buf, 0);
+            assert(len == bufNeeded);
+        }
+        catch (ShortBufferException e)
+        {
+            /* Shouldn't happen, we allocated enough. */
+            throw new IllegalStateException(e);
+        }
+        return buf;
+    }
+
+    /**
+     * Call EVP_CipherUpdate, throwing on failure.
+     */
+    private void doCipherUpdate(byte[] input, int inputOffset, int inputLen,
+        byte[] output, int outputOffset)
+    {
+        if (!EVP_CipherUpdate(ctx, input, inputOffset, inputLen, output,
+            outputOffset))
+        {
+            throw new IllegalStateException("Failure in EVP_CipherUpdate");
+        }
+    }
+
+    private int doGcmDecryptUpdateWithBuffer(byte[] input,
+        int inputOffset, int inputLen,
+        byte[] output, int outputOffset)
+    {
+        /* For GCM decryption, we need to hold on to the bytes that might be
+         * the auth tag. */
+        if (buffer == null)
+        {
+            buffer = new byte[tagLen];
+        }
+        int len = buffered + inputLen - tagLen;
+        int outLen = 0;
+        if (len > 0)
+        {
+            if (len <= buffered)
+            {
+                doCipherUpdate(buffer, 0, len, output, outputOffset);
+                buffered -= len;
+                inputOffset += len;
+                inputLen -= len;
+                if (buffered > 0)
+                {
+                    System.arraycopy(buffer, len, buffer, 0, buffered);
+                }
+                outLen += len;
+            }
+            else
+            {
+                if (buffered > 0)
+                {
+                    doCipherUpdate(buffer, 0, buffered, output,
+                        outputOffset);
+                    outputOffset += buffered;
+                    len -= buffered;
+                    buffered = 0;
+                    outLen += buffered;
+                }
+                if (len > 0)
+                {
+                    doCipherUpdate(input, inputLen, len, output,
```

Comment:
I'm a bit confused about this. If the method is called the first time from engineUpdate (i.e. buffered is still zero) for a block that is e.g. 28bytes (1 block of AES=16bytes + 12bytes tag), wouldn't this just update the cipher with _only_ the tag? 

Also, as you said the CipherSpi could theoretically be reused after calling egineInit, `buffer` should be cleared and `buffered` reset.